### PR TITLE
[Confidential Ledger][Data Plane] Introduce new API version 2023-01-18-preview

### LINKED ISSUE #21804: [Azure Confidential Computing Ledger] API Review
New API Review meeting has been requested.

**Service Name**: Azure Confidential Computing Ledger
**Review Created By**: Andrea Piccione @andpiccione
**Review Date**: 1/4/2023 9:00 AM PT
**PR**: https://github.com/Azure/azure-rest-api-specs/pull/21659
**Hero Scenarios Link**: Not Provided
**Core Concepts Doc Link**: Not Provided

**Description**: See PR description: https://github.com/Azure/azure-rest-api-specs/pull/21659#issue-1464593680

Detailed meeting information and documents provided can be accessed [here](https://web.powerapps.com/apps/e41c69de-ef96-44ba-b287-0e012e7c13ae?view=edit&event=48ef411c-3a76-ed11-81ab-000d3a341bd5)
For more information that will help prepare you for this review, the requirements, and office hours, visit the documentation [here](https://aka.ms/azsdk/onboarding/restapischedule)

### PR DESCRIPTION
**Pull request by andpiccione**
# Data Plane API - Pull Request

This PR adds new Data Plane API specifications for Azure Confidential Ledger. The new API version (`2023-01-18-preview`) adds a new  `applicationClaims` object field to [the response payload of the write transaction receipt](https://github.com/Azure/azure-rest-api-specs/blob/752cf2f8ed18982c515a8693f01e54da08cb14eb/specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json#L902) returned by a Confidential Ledger resource through the [`GetReceipt` operation](https://github.com/Azure/azure-rest-api-specs/blob/752cf2f8ed18982c515a8693f01e54da08cb14eb/specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json#L294).

For context, _application claims_ are arbitrary data that can be attached to write transactions by an Azure Confidential Ledger application, and they effectively represent statements of what was executed during a write operation. When application claims are attached to a transaction, the SHA-256 digest of the claims object is appended to the ledger and committed as part of the write transaction itself (thus, the digest is signed in place and cannot be tampered with). 

Application claims can later be revealed in their "plain" / un-digested form in the [receipt](https://learn.microsoft.com/en-us/azure/confidential-ledger/write-transaction-receipts) payload corresponding to the same transaction where claims have been added. Users can leverage the information attached to the receipt to "re-compute" the same claims digest that was attached and signed in place by the Azure Confidential Ledger instance during the transaction. The claims digest can be used as part of the [write transaction receipt verification process](https://learn.microsoft.com/en-us/azure/confidential-ledger/verify-write-transaction-receipts) and allows users to fully verify offline the authenticity of the recorded claims.

More details about application claims can also be found at the following CCF documentation links:
- [Application Claims](https://microsoft.github.io/CCF/main/use_apps/verify_tx.html#application-claims)
- [User-Defined Claims in Receipts](https://microsoft.github.io/CCF/main/build_apps/example_cpp.html#user-defined-claims-in-receipts)

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*engagement record*](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/271/Azure-REST-API-Stewardship?anchor=rest-api-stewardship-process).

* Link to engagement record issue: N/A (let me know if I need to create a GitHub issue but I think it shouldn't be needed)

Is this review for (select one):

- [x] a private preview
- [ ] a public preview
- [ ] GA release

### Change Scope
<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous Open API document (swagger) if applicable, and the root paths that have been updated. </sup>

* Design Document: [Application claims in CCF](https://microsoft.github.io/CCF/main/use_apps/verify_tx.html?transactions=#application-claims)
* Previous Open API Doc: [2022-05-13 specs](https://github.com/Azure/azure-rest-api-specs/tree/main/specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-05-13)
* Updated paths: 
  * [Path and operation affected](https://github.com/Azure/azure-rest-api-specs/blob/752cf2f8ed18982c515a8693f01e54da08cb14eb/specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json#L292-L324) (only the response payload was updated)
  * [New field in receipt payload response](https://github.com/Azure/azure-rest-api-specs/blob/752cf2f8ed18982c515a8693f01e54da08cb14eb/specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json#L909)
  * [New definitions](https://github.com/Azure/azure-rest-api-specs/blob/752cf2f8ed18982c515a8693f01e54da08cb14eb/specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json#L493-L530)
  * [Updated GetReceipt example](https://github.com/Azure/azure-rest-api-specs/blob/752cf2f8ed18982c515a8693f01e54da08cb14eb/specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/examples/ledger/GetReceipt.json) and [new field example](https://github.com/Azure/azure-rest-api-specs/blob/752cf2f8ed18982c515a8693f01e54da08cb14eb/specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/examples/ledger/GetReceipt.json#L10-L19)

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azapi) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azapi).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling
 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications
 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links
 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

fix https://github.com/Azure/azure-rest-api-specs/issues/21804

### CODE REVIEW

File: specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json
Comment (andpiccione) [inline specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json:963]: The `Breaking Change(Cross-Version)` CI check is failing due to [a runtime exception](https://github.com/Azure/azure-rest-api-specs/pull/21659/checks?check_run_id=9712102995) when parsing this value. The value wasn't changed in the new API version and it is exactly the same as the previous one (see [here](https://github.com/Azure/azure-rest-api-specs/blob/1fefe3f5cee88319b17c08a2dbf95e1e983a9f8c/specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-05-13/confidentialledger.json#L922) for comparison). I raised an issue in the `openapi-diff` repository: https://github.com/Azure/openapi-diff/issues/258.

File: specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json
Comment (andpiccione) [reply specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json:9]: I modified these fields explicitly because the CI was failing the linter check because of this (see [this rule](https://github.com/Azure/azure-openapi-validator/blob/main/docs/host-parameters-validation.md)). I'd check with the API board / reviewers if we need to change or suppress this rule somehow.

File: specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json
Comment (andpiccione) [reply specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json:514]: As also mentioned in another [comment](https://github.com/microsoft/TPAL/pull/214#discussion_r1040070507), I'm not convinced that we should commit to this structure right now, given the lengthy discussion we had to decide how to reveal application claims in a receipt. 

There is the possibility that we decide to change again the application claims structure and converting properties from `required` to `optional`, or viceversa, is considered a breaking change for an Azure API. Therefore, I'd prefer to keep the structure more flexible to make future changes easier to carry out. But please let me know if you have a different opinion and we can discuss about it!

> I also seem to remember that `claimData` may not be a full JSON object if we are including the claim but not revealing it. If that is the case, please express it in the model as well.

Nope, `claimData` is supposed to always be a full JSON object. If we don't reveal the claim, we use a different field (`claimDigest`) to express the unrevealed claim (I'm adding it now to the model as I forgot to add it before 😅)

File: specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json
Comment (andpiccione) [reply specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json:913]: Yes, I thought about it when creating application claims and I was initially considering having a `nextLink` useless given that we currently support one write per transaction (and thus only one claim) and support for multiple writes during a transaction is probably not in scope in the near feature. However, we do not have any visibility into how many application claims we may be able to attach to a receipt in the future (even if we only stick to 1 write per transaction), so having a `nextLink` could actually make sense for extensibility purposes.

Technically, the same reasoning should apply on the arrays inside the receipt content (i.e., `proof` and `serviceEndorsements`), but I'm not sure if there was some reason to not have a `nextLink` there, do you happen to know?

I'll probably wait for the API board review before adding a possible `nextLink` field to discuss the use case. However, in case we agree to add it, that would imply we should have the `applicationClaims` array field as a `required` property, while we decided long time ago that it should rather be `optional`. To mitigate, we could potentially add another top structure, such as:

```
{
  "receipt": {
    ...
  },
  "state": "Ready",
  "transactionId": "2.40",
  "applicationClaims": {
    "claims": [
      {
        "claimData": {
          "collectionId": "My collection ID",
          "contents": "My message"
        },
        "secretKey": "Y2Fhc2RzZGFzZGFzZGFzZHNhdmRoYXNqZGhhc2tqZGhhcw==",
        "version": "v1"
      },
      {
        "claimDigest": "221627395c79fc4ffd8e4c2f139bd99aa83c4aa6bde83da39516a1f6bc830667",
        "version": "v1"
      }
    ],
    "nextLink": ""
  }
}
``` 
Let me know if you have any thoughts on this!

**Comment by andpiccione** [comment]
> Hi, @andpiccione, For review efficiency consideration, when creating a new api version, it is required to place API specs of the base version in the first commit, and push new version updates into successive commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version). Or you could onboard [API spec pipeline](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/606/Enable-API-Spec-Pipeline)

My first commit included the base API specs version, with just a change to the version number. Hopefully this should suffice.

File: specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json
Comment (lynshi) [reply specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json:514]: Sorry, I didn't fully think through which fields need to be required, just mentioned it for you to think about. I think we want `version` to be required. Then if we have a version 2 that uses new fields, we can add those additional fields as optional without removing the old fields. But the `version` parameter is super important and makes sense to be required because you cannot do any verification without it, as `version` exactly specifies the fields required in the claim as well as the methodology for creating the claim.

And yeah you're right, `claimData` / `claimDigest` are distinct, but you were definitely missing one so glad you got the intent 😆

File: specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json
Comment (lynshi) [reply specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json:913]: It was perfectly ok to say a list probably doesn't need to be paged. For example, proof and endorsements are pretty bounded, and application claims makes sense to be bounded as well. Additionally, the receipt is intended to be downloaded all at once, so there is no client-side benefit of "paging over" receipt contents as there might be in a normal collection download where processing of N individual elements is supposed to happen separately (e.g. if you are looping over a collection, you can start processing each item as you receive them instead of waiting up front for all items to be downloaded).

Sorry for the wild goose chase! I think it's unlikely we get asked about it anyway as this list is more buried and is not a top-level collection return. I just wanted to let you know about the possibility so you can be prepared to answer if needed.

File: specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json
Comment (andpiccione) [reply specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json:514]: I understand that the `version` field is important but I don't feel like we are fully established on the current structure, and I fear making some fields required now would put some constraints on how we can evolve the model in the future. However, it's also true that the list structure wrapping `ApplicationClaimObjects` is optional too, so changing that later on should not be a big problem theoretically. Based on this, considering that `claimData` / `claimDigest` may be more prone to change (due to the nature of the data we may want to use as claims), I'd say `version` and `secretKey` can both be required properties as you were suggesting in your initial comment.

File: specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json
Comment (andpiccione) [reply specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json:913]: I'm not too sold on this :D I agree that a receipt is supposed to be downloaded all at once (and making it paginated would not be great for sure), but at the moment we don't know how many claims we could support in the future for a single transaction. In theory, they should be bounded to something (e.g., the number of entries written in the ledger table during a transaction), but in practice we don't have such limit in place (and we don't have a way to derive a reasonable limit to use), so technically nothing prevents a Ledger from returning N claims, where N can be very big, due to a bug or a logical error.

I see this is likely a very small problem and 99% we should be fine with the current approach. But from an API model point of view, maybe it's worth understanding if having a `nextLink` is worth in this scenario, hence my second thoughts in my previous comment. Let's discuss this during the API review and see what the board suggests.

File: specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json
Comment (andpiccione) [reply specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json:514]: Addressed in https://github.com/Azure/azure-rest-api-specs/pull/21659/commits/45629addd0a0f6ce4ec567f9eb98eee4adaa7c74

**Review by lynshi — APPROVED** [review]

**Review by markweitzel — CHANGES_REQUESTED** [review]
See discussion & next steps requested from API Stewardship Board captured via #21804

File: specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json
Comment (andpiccione) [reply specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json]: We decided that we don't need pagination for application claims as they will be limited in size.

**Review by lynshi — APPROVED** [review]

**Review by mikekistler — APPROVED** [review]
Looks good! 👍

**Comment by mikekistler** [comment]
@JeffreyRichter @johanste @tg-msft Could we get one more review/approval on this PR so the service team can move forward on this?

**Review by JeffreyRichter — APPROVED** [review]

File: specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json
Comment (andpiccione) [reply specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json]: @mikekistler @markweitzel @JeffreyRichter @johanste @lynshi Do you have any opinion on the above problem and how we could solve it?

File: specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json
Comment (andpiccione) [reply specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json]: @mikekistler @markweitzel @JeffreyRichter @johanste @lynshi Do you have any opinion on this? Should we suppress this rule or comply with the check?

**Comment by andpiccione** [comment]
@markweitzel Could you also please review the changes as well since you requested changes and the merge is blocked by that?

File: specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json
Comment (JeffreyRichter) [reply specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json]: I'm not familiar with this rule at all. @johanste will have to answer

**Comment by mikekistler** [comment]
@jhendrixMSFT This PR is approved and ready to merge.

**Comment by andpiccione** [comment]
@mikekistler Thanks for pushing this through. Before merging, we would like to just resolve the last two unresolved comments above and make the required changes in our backend to comply with the latest model changes. We will drop a message here when we are ready to merge.

File: specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json
Comment (andpiccione) [reply specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json]: I believe it should not be a problem to reduce the number (@lynshi please correct me otherwise). I tried changing the limits to the respective maximum (2^53 – 1) and minimum (-(2^53 – 1)) safe integers in JavaScript, but we are still getting the same runtime exception. If that is ok for the board, we can ignore this error and merge the PR anyway since this is a non-issue in my opinion.

File: specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json
Comment (andpiccione) [reply specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json]: I decided to revert this change back and use the same naming of the previous API versions for compatibility reasons. Hopefully, it should be fine to ignore the linter rule this time given the circumstances. Feel free to let me know if you have a different opinion or a better suggestion.

**Comment by andpiccione** [comment]
@jhendrixMSFT The PR is _almost_ ready to be merged (just waiting for a few changes to be finalized on our backend, I'll drop a message here when they are merged so we can close this too), but I just want to make sure there are no other blockers related to the CI checks failing. The failures have been documented in the PR comments above, please let me know if there is anything else that needs to be addressed or if there is some other workaround for these. Thanks!

File: specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json
Comment (jhendrixMSFT) [reply specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json]: IIRC, this rule was due to autorest and/or code generators making an assumption about the name of the endpoint parameter beinig `endpoint`.  Do we know if such an assumption remains?  The above alludes to it working with a different name.  If this requirement is no longer necessary then we should remove this linter rule.

Comment (jhendrixMSFT) [reply specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json]: @lmazuel if this is ok with you can you please merge?

**Comment by andpiccione** [comment]
@jhendrixMSFT @lmazuel The PR is ready to be merged.

File: specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json
Comment (andpiccione) [reply specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json]: Closing as no alternative solution was provided for fixing this CI check.

**Comment by lmazuel** [comment]
Merging, since I see approval from @mikekistler and @JeffreyRichter after the request changes from @markweitzel , so I assume it's ok to merge.

**Pull request merged**
